tools/opensnoop.py

```diff
@@ -137,38 +181,28 @@
 
     return 0;
 };
+"""
 
-int trace_return(struct pt_regs *ctx)
+bpf_text_kfunc_header_open = """
+KRETFUNC_PROBE(FNNAME, struct pt_regs *regs, int ret)
 {
-    u64 id = bpf_get_current_pid_tgid();
-    struct val_t *valp;
-    struct data_t data = {};
-
-    u64 tsp = bpf_ktime_get_ns();
-
-    valp = infotmp.lookup(&id);
-    if (valp == 0) {
-        // missed entry
-        return 0;
-    }
-    bpf_probe_read_kernel(&data.comm, sizeof(data.comm), valp->comm);
-    bpf_probe_read_user(&data.fname, sizeof(data.fname), (void *)valp->fname);
-    data.id = valp->id;
-    data.ts = tsp / 1000;
-    data.uid = bpf_get_current_uid_gid();
-    data.flags = valp->flags; // EXTENDED_STRUCT_MEMBER
-    data.ret = PT_REGS_RC(ctx);
-
-    events.perf_submit(ctx, &data, sizeof(data));
-    infotmp.delete(&id);
+    const char __user *filename = (char *)regs->di; int flags = regs->si;
```

Comment:
The same as below, we cannot use x86 specific registers. This will make tool not portable. Also, `regs` cannot be used to access all registers. You need to do something like below for x64 and arm64.
```
#if defined(CONFIG_ARCH_HAS_SYSCALL_WRAPPER) && !defined(__s390x__)
   struct pt_regs *real_regs = PT_REGS_PARM1(regs);
#else
   struct pt_regs *real_regs = regs;
#endif
   const char __user *filename = PT_REGS_PARAM1(real_regs);
   int flags = PT_REGS_PARAM2(real_regs);
   ...
```

It would be good if we can hide the above macros from user. But this can be delayed until future work.

Reply:
> ```
> #if defined(CONFIG_ARCH_HAS_SYSCALL_WRAPPER) && !defined(__s390x__)
>    struct pt_regs *real_regs = PT_REGS_PARM1(regs);
> #else
>    struct pt_regs *real_regs = regs;
> #endif
>    const char __user *filename = PT_REGS_PARAM1(real_regs);
>    int flags = PT_REGS_PARAM2(real_regs);
>    ...
> ```
This only works for kprobes which collects parameter(s) as pt_regs and passes it to the registered hook, so we need to extract twice for SYSCALL_WRAPPER.

For kfunc, we have direct access to the parameter(s). Do you mean the following definition?
```
#if defined(CONFIG_ARCH_HAS_SYSCALL_WRAPPER) && !defined(__s390x__)
KRETFUNC_PROBE(FNNAME, struct pt_regs *regs, int ret)
{
    const char __user *filename = (char *)PT_REGS_PARM1(regs);
    int flags = PT_REGS_PARM2(regs);
#else
KRETFUNC_PROBE(FNNAME, const char __user *filename, int flags, int ret)
{
#endif
```